include/dlaf/cuda/pool.h

```diff
@@ -0,0 +1,131 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+/// @file
+
+#include <vector>
+
+#include <cublas_v2.h>
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+
+#include "dlaf/common/assert.h"
+#include "dlaf/cuda/error.h"
+
+namespace dlaf {
+
+// Print information about the devices.
+inline void print_device_info(int device) noexcept {
+  cudaDeviceProp device_prop;
+  DLAF_CUDA_CALL(cudaGetDeviceProperties(&device_prop, device));
+  printf("Device %d has compute capability %d.%d.\n", device, device_prop.major, device_prop.minor);
+}
+
+// Return the number of devices.
+inline int num_devices() noexcept {
+  int ndevices;
+  cudaGetDeviceCount(&ndevices);
+  return ndevices;
+}
+
+// A pool of handles and streams assigned to a `device`.
+//
+// To amortize creation/destruction costs with CUBLAS handles/streams, the recommended best practice is
+// to create all the needed CUBLAS handles/streams upfront and destroy them after the work is done [1].
+//
+// [1]: https://github.com/pytorch/pytorch/issues/9646
+class cublas_pool {
+  int device_;
+  std::vector<cublasHandle_t> handles_arr_;
+  int curr_handle_id_;
+
+public:
+  cublas_pool(cublas_pool&&) = default;
+  cublas_pool& operator=(cublas_pool&&) = default;
+  cublas_pool(const cublas_pool&) = delete;
+  cublas_pool& operator=(const cublas_pool&) = delete;
+
+  /// The number of streams spawned per device depends on the supported maximum number of concurrent
+  /// kernels that can execute in parallel on that device [1]. The number varies between different GPU
+  /// models, most support at least 16.
+  ///
+  /// [1]: CUDA Programming Guide, Table 15. Technical Specifications per Compute Capability
+  cublas_pool(int device, int num_streams) noexcept : device_{device}, curr_handle_id_{0} {
+    DLAF_ASSERT(num_streams >= 1, "At least 1 stream needs to be specified!");
+    DLAF_ASSERT(device >= 0, "Valid device IDs start from 0!");
+    DLAF_ASSERT(device < num_devices(), "The device ID exceeds the number of available devices!");
+
+    handles_arr_.reserve(static_cast<std::size_t>(num_streams));
+
+    // - A kernel launch will fail if it is issued to a stream that is not associated to the current
+    //   device. [1]
+    //
+    // [1]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#stream-and-event-behavior
+    cudaSetDevice(device);
+    for (int i_stream = 0; i_stream < num_streams; ++i_stream) {
+      cudaStream_t stream;
+      DLAF_CUDA_CALL(cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking));
+      cublasHandle_t handle;
+      DLAF_CUBLAS_CALL(cublasCreate(&handle));
+      DLAF_CUBLAS_CALL(cublasSetStream(handle, stream));
+      handles_arr_.push_back(handle);
```

Comment:
Having a separate handle per stream avoids this:

> - set handle stream

which I think is a win.

Reply:
cublasSetStream adds no overhead; it is just setting the id of stream that will be used for next cublas calls.
On the other hand "cublasCreate allocates some internal resources" https://docs.nvidia.com/cuda/cublas/index.html#cublascreate

Reply:
`cublasCreate()` is only called at the creation of the pool. Usually, a single pool suffices for the entire duration of the program and can be used by multiple executors. We are essentially trading, memory overhead for some execution (perhaps minimal) overhead. I haven't benchmarked `cublasSetStream()` and I am not confident whether it "adds no overhead"... I'd expect there is at least an atomic operation or some mutex involved because multiple threads may call the function.